exonum/src/blockchain/transaction.rs

```diff
@@ -0,0 +1,398 @@
+// Copyright 2017 The Exonum Team
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! `Transaction` related types.
+
+use std::borrow::Cow;
+use std::fmt;
+use std::u8;
+
+use messages::Message;
+use storage::{Fork, StorageValue};
+use crypto::{Hash, CryptoHash};
+use encoding::serialize::json::ExonumJson;
+
+//  User-defined error codes (`TransactionError::Code(u8)`) have a `0...255` range.
+#[cfg_attr(feature = "cargo-clippy", allow(cast_lossless))]
+const MAX_ERROR_CODE: u16 = u8::MAX as u16;
+// Represent `(Ok())` `TransactionStatus` value.
+const TRANSACTION_STATUS_OK: u16 = MAX_ERROR_CODE + 1;
+// `Err(TransactionError::Panic)`.
+const TRANSACTION_STATUS_PANIC: u16 = TRANSACTION_STATUS_OK + 1;
+// `Err(TransactionError::Failure)`.
+const TRANSACTION_STATUS_FAILURE: u16 = TRANSACTION_STATUS_PANIC + 1;
+// `Err(TransactionError::Description)`.
+const TRANSACTION_STATUS_DESCRIPTION: u16 = TRANSACTION_STATUS_FAILURE + 1;
+
+/// Return value of the `Transaction`'s `execute' method. Changes made by the transaction are
+/// discarded if `Err` is returned, see `Transaction` documentation for the details.
+pub type ExecutionStatus = Result<(), ExecutionError>;
```

Comment:
It's a bit strange that `Result` type does not have "Result" in the name.

Reply:
I'm not so sure myself. On the one hand, I simply don't like such typedefs: I prefer to have `Result` in the appropriate namespace. On the other hand, forcing user to write something like `use exonum::transaction::Result as TransactionResult` isn't good either.

We need to seriously revise our imports anyway, but it can happen not so soon.

Lets use your comment as a little poll: :+1: for `Result` and :-1: for `Status`.